In [16]:
!pip install ipython-sql 
%load_ext sql
%sql sqlite://

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [2]:
%%sql --copied and pasted the Schema SQL code to create permanent tables
CREATE TABLE runners (
  "runner_id" INTEGER,
  "registration_date" DATE
);
INSERT INTO runners
  ("runner_id", "registration_date")
VALUES
  (1, '2021-01-01'),
  (2, '2021-01-03'),
  (3, '2021-01-08'),
  (4, '2021-01-15');


DROP TABLE IF EXISTS customer_orders;
CREATE TABLE customer_orders (
  "order_id" INTEGER,
  "customer_id" INTEGER,
  "pizza_id" INTEGER,
  "exclusions" VARCHAR(4),
  "extras" VARCHAR(4),
  "order_time" TIMESTAMP
);

INSERT INTO customer_orders
  ("order_id", "customer_id", "pizza_id", "exclusions", "extras", "order_time")
VALUES
  ('1', '101', '1', '', '', '2020-01-01 18:05:02'),
  ('2', '101', '1', '', '', '2020-01-01 19:00:52'),
  ('3', '102', '1', '', '', '2020-01-02 23:51:23'),
  ('3', '102', '2', '', NULL, '2020-01-02 23:51:23'),
  ('4', '103', '1', '4', '', '2020-01-04 13:23:46'),
  ('4', '103', '1', '4', '', '2020-01-04 13:23:46'),
  ('4', '103', '2', '4', '', '2020-01-04 13:23:46'),
  ('5', '104', '1', 'null', '1', '2020-01-08 21:00:29'),
  ('6', '101', '2', 'null', 'null', '2020-01-08 21:03:13'),
  ('7', '105', '2', 'null', '1', '2020-01-08 21:20:29'),
  ('8', '102', '1', 'null', 'null', '2020-01-09 23:54:33'),
  ('9', '103', '1', '4', '1, 5', '2020-01-10 11:22:59'),
  ('10', '104', '1', 'null', 'null', '2020-01-11 18:34:49'),
  ('10', '104', '1', '2, 6', '1, 4', '2020-01-11 18:34:49');


DROP TABLE IF EXISTS runner_orders;
CREATE TABLE runner_orders (
  "order_id" INTEGER,
  "runner_id" INTEGER,
  "pickup_time" VARCHAR(19),
  "distance" VARCHAR(7),
  "duration" VARCHAR(10),
  "cancellation" VARCHAR(23)
);

INSERT INTO runner_orders
  ("order_id", "runner_id", "pickup_time", "distance", "duration", "cancellation")
VALUES
  ('1', '1', '2020-01-01 18:15:34', '20km', '32 minutes', ''),
  ('2', '1', '2020-01-01 19:10:54', '20km', '27 minutes', ''),
  ('3', '1', '2020-01-03 00:12:37', '13.4km', '20 mins', NULL),
  ('4', '2', '2020-01-04 13:53:03', '23.4', '40', NULL),
  ('5', '3', '2020-01-08 21:10:57', '10', '15', NULL),
  ('6', '3', 'null', 'null', 'null', 'Restaurant Cancellation'),
  ('7', '2', '2020-01-08 21:30:45', '25km', '25mins', 'null'),
  ('8', '2', '2020-01-10 00:15:02', '23.4 km', '15 minute', 'null'),
  ('9', '2', 'null', 'null', 'null', 'Customer Cancellation'),
  ('10', '1', '2020-01-11 18:50:20', '10km', '10minutes', 'null');


DROP TABLE IF EXISTS pizza_names;
CREATE TABLE pizza_names (
  "pizza_id" INTEGER,
  "pizza_name" TEXT
);
INSERT INTO pizza_names
  ("pizza_id", "pizza_name")
VALUES
  (1, 'Meatlovers'),
  (2, 'Vegetarian');


DROP TABLE IF EXISTS pizza_recipes;
CREATE TABLE pizza_recipes (
  "pizza_id" INTEGER,
  "toppings" TEXT
);
INSERT INTO pizza_recipes
  ("pizza_id", "toppings")
VALUES
  (1, '1, 2, 3, 4, 5, 6, 8, 10'),
  (2, '4, 6, 7, 9, 11, 12');


DROP TABLE IF EXISTS pizza_toppings;
CREATE TABLE pizza_toppings (
  "topping_id" INTEGER,
  "topping_name" TEXT
);
INSERT INTO pizza_toppings
  ("topping_id", "topping_name")
VALUES
  (1, 'Bacon'),
  (2, 'BBQ Sauce'),
  (3, 'Beef'),
  (4, 'Cheese'),
  (5, 'Chicken'),
  (6, 'Mushrooms'),
  (7, 'Onions'),
  (8, 'Pepperoni'),
  (9, 'Peppers'),
  (10, 'Salami'),
  (11, 'Tomatoes'),
  (12, 'Tomato Sauce');

 * sqlite://
Done.
4 rows affected.
Done.
Done.
14 rows affected.
Done.
Done.
10 rows affected.
Done.
Done.
2 rows affected.
Done.
Done.
2 rows affected.
Done.
Done.
12 rows affected.


[]

In [3]:
%%sql --checking data to ensure that it was loaded correctly

SELECT * from runners

 * sqlite://
Done.


[(1, '2021-01-01'), (2, '2021-01-03'), (3, '2021-01-08'), (4, '2021-01-15')]

In [4]:
%%sql --review runner_orders table
SELECT *
FROM runner_orders

 * sqlite://
Done.


[(1, 1, '2020-01-01 18:15:34', '20km', '32 minutes', ''),
 (2, 1, '2020-01-01 19:10:54', '20km', '27 minutes', ''),
 (3, 1, '2020-01-03 00:12:37', '13.4km', '20 mins', None),
 (4, 2, '2020-01-04 13:53:03', '23.4', '40', None),
 (5, 3, '2020-01-08 21:10:57', '10', '15', None),
 (6, 3, 'null', 'null', 'null', 'Restaurant Cancellation'),
 (7, 2, '2020-01-08 21:30:45', '25km', '25mins', 'null'),
 (8, 2, '2020-01-10 00:15:02', '23.4 km', '15 minute', 'null'),
 (9, 2, 'null', 'null', 'null', 'Customer Cancellation'),
 (10, 1, '2020-01-11 18:50:20', '10km', '10minutes', 'null')]

In [5]:
%%sql --review customer_orders table
SELECT *
FROM customer_orders

 * sqlite://
Done.


[(1, 101, 1, '', '', '2020-01-01 18:05:02'),
 (2, 101, 1, '', '', '2020-01-01 19:00:52'),
 (3, 102, 1, '', '', '2020-01-02 23:51:23'),
 (3, 102, 2, '', None, '2020-01-02 23:51:23'),
 (4, 103, 1, '4', '', '2020-01-04 13:23:46'),
 (4, 103, 1, '4', '', '2020-01-04 13:23:46'),
 (4, 103, 2, '4', '', '2020-01-04 13:23:46'),
 (5, 104, 1, 'null', '1', '2020-01-08 21:00:29'),
 (6, 101, 2, 'null', 'null', '2020-01-08 21:03:13'),
 (7, 105, 2, 'null', '1', '2020-01-08 21:20:29'),
 (8, 102, 1, 'null', 'null', '2020-01-09 23:54:33'),
 (9, 103, 1, '4', '1, 5', '2020-01-10 11:22:59'),
 (10, 104, 1, 'null', 'null', '2020-01-11 18:34:49'),
 (10, 104, 1, '2, 6', '1, 4', '2020-01-11 18:34:49')]

# Question 1: 
How many runners signed up for each 1 week period? (i.e. week starts 2021-01-01)

In [6]:
%%sql
SELECT COUNT(runner_id) as total_runners_registered, strftime('%Y-%W', registration_date) AS week_of_registration
FROM runners
GROUP BY week_of_registration
ORDER BY week_of_registration

 * sqlite://
Done.


[(2, '2021-00'), (1, '2021-01'), (1, '2021-02')]

In the week_of_registration column, the number to the right of the dash represents the week of the year, with the first week containing a Monday counted as week 1. So, to answer this question, it appears that two runners registered during the first week in 2021 (short week), one runner registered in week 2, and one runner registered in week 3. 

# Question 2: 
What was the average time in minutes it took for each runner to arrive at the Pizza Runner HQ to pickup the order?

In [7]:
%%sql
/* Create a temporary TIMESTAMP column*/
ALTER TABLE runner_orders ADD COLUMN updated_pickup_time TIMESTAMP without time zone NULL;

/* Copy casted value over to the temporary column */
UPDATE runner_orders SET updated_pickup_time = pickup_time;

 * sqlite://
Done.
Done.


[]

In [8]:
%%sql
/* Create a temporary TIMESTAMP column */
ALTER TABLE customer_orders ADD COLUMN updated_order_time TIMESTAMP without time zone NULL;

/* Copy casted value over to the temporary column */
UPDATE customer_orders SET updated_order_time = order_time;

 * sqlite://
Done.
Done.


[]

In [9]:
%%sql
PRAGMA table_info(customer_orders);

 * sqlite://
Done.


[(0, 'order_id', 'INTEGER', 0, None, 0),
 (1, 'customer_id', 'INTEGER', 0, None, 0),
 (2, 'pizza_id', 'INTEGER', 0, None, 0),
 (3, 'exclusions', 'VARCHAR(4)', 0, None, 0),
 (4, 'extras', 'VARCHAR(4)', 0, None, 0),
 (5, 'order_time', 'TIMESTAMP', 0, None, 0),
 (6, 'updated_order_time', 'TIMESTAMP without time zone', 0, None, 0)]

In [10]:
%%sql
SELECT runner_id, ROUND(AVG(julianday(updated_pickup_time)-julianday(updated_order_time))*24*60) as average_arrival_time
FROM runner_orders
JOIN customer_orders on runner_orders.order_id=customer_orders.order_id
WHERE updated_pickup_time != 'null'
GROUP BY runner_id;

 * sqlite://
Done.


[(1, 16.0), (2, 24.0), (3, 10.0)]

On average, it took runner #1 16 minutes, runner #2 took 24 minutes, and runner #3 took 10 minutes to arrive at the Pizza Runner HQ to pick up their orders

# Question 3:
What was the average distance travelled for each customer?

In [12]:
%%sql
SELECT AVG(TRIM(distance, 'km')) as average_distance_traveled
FROM runner_orders;

 * sqlite://
Done.


[(14.52,)]

The average distance traveled was 14.52 kilometers

# Question 4:
What was the difference between the longest and shortest delivery times for all orders?

In [13]:
%%sql
SELECT MAX(TRIM(duration, 'minutes'))-MIN(TRIM(duration, 'minutes')) as difference_in_delivery_times 
FROM runner_orders
WHERE duration != 'null';

 * sqlite://
Done.


[(30,)]

The difference in delivery times is 30 minutes

# Question 5:
What was the average speed for each runner for each delivery and do you notice any trend for these values?

In [14]:
%%sql
SELECT runner_id, COUNT(runner_id) as number_of_deliveries, AVG(TRIM(duration, 'minutes')) as average_speed
FROM runner_orders
WHERE duration != 'null'
GROUP BY runner_id
ORDER BY average_speed DESC;

 * sqlite://
Done.


[(2, 3, 26.666666666666668), (1, 4, 22.25), (3, 1, 15.0)]

It appears that runner #3 had the least number of deliveries and the shortest average delivery speed. The other runners have significantly higher average speed times (26.67 and 22.25 mins) and relatively similar number of deliveries (3 and 4). One thing to note: Pizza Runner HQs had a total of 4 runners registered, but only three managed to be available for work.